### ISSUE TITLE
[vm_service] Add the ability to use "await" in eval expressions

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#44999]
At the moment `VmService.evaluate` does not support awaiting Futures.

It would be useful to extend `evaluate` to allow:

```
await Future(..);
```
as expression

**Comment by jacob314** [comment]
Fyi @bkonyi 
This would handy in a range of cases in the inspector and debugger.

**Comment by lrhn** [comment]
You'd have to define the intended behavior, which isn't obvious. 
With synchronous computation, the only thing to happen is evaluation of the expression you wrote. It runs, then it's done, and you're back at the breakpoint where you stopped the program.

If you allow asynchronous computation in the debugger, then you need to run an event loop in order to ensure that the expression completes (well, if it ever completes, you should also prepare for the possibility that you are awaiting a future which never completes).

To get to the event loop, you have to let the currently blocked-at-breakpoint code continue running. That means that when the `await` completes (again, if it does), then you're no longer at the same breakpoint. Control could be anywhere, theoretically. In practice control is almost certainly in the middle of the future result propagation loop, which is not a place that is expecting to be interrupted! 
The expression can continue evaluating anyway, by capturing the local variables from when you started the evaluation like a closure would.

**Comment by mraleph** [comment]
I thought about this a bit in context of @annagrin's duplicate request (https://github.com/dart-lang/sdk/issues/45454). 

This made me realize that we could theoretically support this specifically if top stack frame is a resumption of an `async/async*` function, because these functions are expecting to be suspended. 

So user attempts to evaluate `await foo` we can simply suspend top stack frame and resume execution until `foo` completes, after which we reconstruct top level frame which was there before and pause at breakpoint again. 

I think that should work - but it has some limitations e.g. you can't evaluate `await` if you are executing synchronous "prologue" of the `async` function, you can't evaluate `await` if top-level frame is a synchronous function, etc. 

Implementation might also be somewhat gnarly because our async/await is desugared into a state machine, rather then implemented on top of something suspendable at arbitrary points like `Fiber`.

**Comment by lrhn** [comment]
If you are inside an `async` or `async*` function, then the stack is probably limited to the current function &mdash; unless you are prior to the first `await` of an `async` function, in which case you're running synchronously on top of the caller. 
I think it would be reasonable to disallow `await` at a breakpoint before the first real `await` of an `async` function.

(You could also use `dart:cli`'s `waitFor` instead of `await` and get effectively the same result anyway, just likely more error-prone).

**Comment by jacob314** [comment]
I agree that using `await` is very difficult when you are paused at a breakpoint. Eval can also be used when not paused at a breakpoint which is actually the case where @rrousselGit was hitting the limitation. It would be nice to support await in that context even it isn't supported when paused.

**Comment by annagrin** [comment]
@jacob314 good point!  @lrhn @mraleph would that be possible?

**Comment by mraleph** [comment]
Yeah, that should be possible. I don't see why it would not work. (Though in VM `_evaluateCompiledExpression` might need to be rewritten. Currently it expects synchronous completion, I think).

**Comment by annagrin** [comment]
We would still need the frontend part of this bug I presume?  Currently `await` expression do not compile. We also would need to see what happens on web side.

[frontend] Expression evaluation fails for await statements with a compilation error #45454

**Comment by annagrin** [comment]
More on the same topic:
https://github.com/dart-lang/sdk/issues/40198
https://github.com/Dart-Code/Dart-Code/issues/2212
